src/schema.rs

```diff
@@ -499,6 +509,112 @@ impl Serialize for RecordField {
     }
 }
 
+/// Parses a **valid** avro schema into the Parsing Canonical Form.
+/// https://avro.apache.org/docs/1.8.2/spec.html#Parsing+Canonical+Form+for+Schemas
+fn parsing_canonical_form(schema: &serde_json::Value) -> String {
+    match schema {
+        serde_json::Value::Number(n) => pcf_float(n),
+        serde_json::Value::Object(map) => pcf_map(map),
+        serde_json::Value::String(s) => pcf_string(s),
+        serde_json::Value::Array(v) => pcf_array(v),
+        _ => unreachable!(),
+    }
+}
+
+fn pcf_map(schema: &Map<String, serde_json::Value>) -> String {
+    use std::borrow::Cow;
+    // Look for the namespace variant up front.
+    let ns = schema.get("namespace").and_then(|v| v.as_str());
+    let mut fields = Vec::new();
+    for (k, v) in schema {
+        // Reduce primitive types to their simple form. ([PRIMITIVE] rule)
+        if schema.len() == 1 && k == "type" {
+            // Invariant: function is only callable from a valid schema, so this is acceptable.
+            if let serde_json::Value::String(s) = v {
```

Comment:
Some (unlikely but still valid) schemas might look like `{"type": {"type": "int"}}` (recursive `"type"`).  I am not sure this case is handled here 🙂

Reply:
I didn't realize that was a valid schema - I'm not sure how that is supposed to be handled in that case. I would think the current implementation would transform it into `{"type":"int"}`, because the `if let Value::String(s) = v { ... }` branch won't be taken and then it will call into `pcf_map` recursively.

I'll do some more thinking and reading on the avro spec.

Reply:
Actually nevermind, it's handled by the schema parsing 😃 (I wrote the code, but already forgot what I did)
https://github.com/flavray/avro-rs/blob/master/src/schema.rs#L281

Disregard this comment, then!